### ISSUE TITLE
[Feature/Bug] Authentication process need to be persistent across downloads

### ISSUE THREAD

**Issue by kopax** [verdaccio/verdaccio#732]
**Describe the bug**

We finally removed `verdaccio-ldap@2.0.0` plugin for some performance and quality reasons.

Between each package download, the authentication plugin is called to authenticate the user.

It should be fine for service such as LDAP that are designed to resist, but it is a bit rough for a service like GitLab. The truth is that  none of them are safe.

Since **we removed the verdaccio-ldap plugin, it use the `htaccess` and we have no issue at all**.

The issue was happening in CI where we use super computer we lots of power. 

We have not been able to mix **ldap** and **htaccess**.

I have made a list of issue impacted by this error down this post.

I think we should wrap the login with a persistent layer like Redis, something fast. It is better than the current pattern which doesn't permit to use verdaccio as a safe npm registry solution, at least with auth plugin installed.

**To Reproduce**

Install any auth plugin.

**Expected behavior**

Persistent session for the user

**Additional context**

- https://github.com/bufferoverflow/verdaccio-gitlab/issues/13
- https://github.com/bufferoverflow/verdaccio-gitlab/issues/4
- https://github.com/Alexandre-io/verdaccio-ldap/issues/33
- https://github.com/verdaccio/verdaccio/issues/541

**Comment by juanpicado** [comment]
Like I mentioned before, `verdaccio` aims to be simple, 0 dependencies. 

**tl;dr**
this has to be done in the plugin(s) itself.

You might use multiple auth plugins and Verdaccio asks serially (in order are defined in the config file) and persist directly in the plugin. `npm` performs decends or thousands of request to the endpoints so, each endpoint must ask to the `auth` plugin whether the origin has access to X resource.

**Comment by kopax** [comment]
I agree with most of it only that I would add it as core feature because of it's criticity and major improvement it can represent for the community. It could still be an option.

Anyhow, what database would you like to see in use for such job?
I'll personally start with a fork of `verdaccio-ldap` and would use a redis store to retrieve the session.

But I am not sure if I understand how to serially use auth plugins (we tried to use LDAP then htpasswd in the same `config.yaml` and both together were not failing, the htpasswd user was considered as a non existing ldap user), if you want to have such plugin separated, I'll appreciate to read the "how to" procedure.

**Comment by juanpicado** [comment]
> It could still be an option.

That's the challenge.  How to make it optional while is compatible with all previous plugins with no breaking changes.

>  Anyhow, what database would you like to see in use for such job?

I'm not backend developer, I have no strong answer to that. 

> I'll personally start with a fork of verdaccio-ldap and would use a redis store to retrieve the session.

You also might *wrap a.k.a extend* such plugin with another plugin implementation to do whatever you want without forking. I believe in separation of concerns.

> I am not sure if I understand how to serially use auth plugins

It's still confusing, many things on verdaccio are still not documented or basically unknown even for me that I work on this (legacy code with 0 history), work in progress.

**Comment by kopax** [comment]
@juanpicado what do you think of : https://github.com/bufferoverflow/verdaccio-gitlab/issues/13#issuecomment-394082427

**Comment by juanpicado** [comment]
I'll wait for @bufferoverflow answer .. I'm curious 🤓

**Comment by kopax** [comment]
He seems to agree for Redis, same here https://github.com/idangozlan/verdaccio-bitbucket/issues/10#issuecomment-394171013 

We are in a middle of a delivery but I will try to have a look at it asap as we need to have ldap at some point. My current idea is to implement `getCache` equivalent that use redis for verdaccio-ldap.

**Comment by juanpicado** [comment]
I'd suggest try `docker pull verdaccio/verdaccio:4.x-next` enabling the new token based on JWT
https://github.com/verdaccio/verdaccio/blob/4.x/docs/config.md#security
that will reduce to the minimum the issue you were reporting here.

**Comment by dpwolfe** [comment]
Link in @juanpicado's post above about JWT usage has moved to: https://github.com/verdaccio/website/blob/master/docs/config.md#security

**Comment by kopax** [comment]
Thanks @dpwolfe , interesting. Hope to see the persistency one day. We still use individual htaccess for our users.

**Comment by kopax** [comment]
No I did not. To be honest, I've been coming and going about 10 times to try to fix that and use the openldap plugin and everytime it was a fail. So I stopped until I got real feedback from people.
Why would `security` feature from `4.x` would solve the problem ? Thanks

**Comment by juanpicado** [comment]
We use JWT and we just verify the token without hit the Auth provider.

**Comment by kopax** [comment]
- Is the JWT personal or system wide?
- Is the JWT the same as the one used in @verdaccio/gitlab-plugin?

**Comment by kopax** [comment]
any news about this? we really need ldap to get working with verdaccio.

**Comment by juanpicado** [comment]
> Is the JWT personal or system wide?

I don't get this one.

> Is the JWT the same as the one used in @verdaccio/gitlab-plugin?

Yes.

> How does the JWT handle group (can he use ldap or gitlab group)?

All groups are packed with the token payload.


@kopax  I think this should be closed it, the original request was persis auth and that cannot be done in the core.

**Comment by kopax** [comment]
I love people who ask to close issue. Yes it is solved with the jwt included in verdaccio v4, thanks a lot for solving this.
